packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -370,59 +362,58 @@ class AndroidDevice extends Device {
+    Uri uri = new Uri(scheme: 'ws', host: '127.0.0.1', port: observatoryPort, path: 'ws');
+    WebSocket ws = await WebSocket.connect(uri.toString());
+    rpc.Client client = new rpc.Client(new IOWebSocketChannel(ws));
+    client.listen();
+    return client;
+  }
 
-    RegExp traceRegExp = new RegExp(r'Saving trace to (\S+)', multiLine: true);
-    RegExp completeRegExp = new RegExp(r'Trace complete', multiLine: true);
-
-    String tracePath;
-    bool isComplete = false;
-    while (!isComplete) {
-      List<String> args = <String>['logcat', '-d'];
-      if (beforeStop != null)
-        args.addAll(<String>['-T', beforeStop]);
-      String logs = runCheckedSync(adbCommandForDevice(args));
-      Match fileMatch = traceRegExp.firstMatch(logs);
-      if (fileMatch != null && fileMatch[1] != null) {
-        tracePath = fileMatch[1];
-      }
-      isComplete = completeRegExp.hasMatch(logs);
+  Future<Null> startTracing(AndroidApk apk, int observatoryPort) async {
+    rpc.Client client;
+    try {
+      client = await _connectToObservatory(observatoryPort);
+    } catch (e) {
+      printError('Error connecting to observatory: $e');
+      return;
     }
 
-    if (tracePath != null) {
-      String localPath = (outPath != null) ? outPath : path.basename(tracePath);
-
-      // Run cat via ADB to print the captured trace file. (adb pull will be unable
-      // to access the file if it does not have root permissions)
-      IOSink catOutput = new File(localPath).openWrite();
-      List<String> catCommand = adbCommandForDevice(
-          <String>['shell', 'run-as', apk.id, 'cat', tracePath]
-      );
-      Process catProcess = await Process.start(catCommand[0],
-          catCommand.getRange(1, catCommand.length).toList());
-      catProcess.stdout.pipe(catOutput);
-      int exitCode = await catProcess.exitCode;
-      if (exitCode != 0)
-        throw 'Error code $exitCode returned when running ${catCommand.join(" ")}';
-
-      runSync(adbCommandForDevice(
-          <String>['shell', 'run-as', apk.id, 'rm', tracePath]
-      ));
-      return localPath;
+    await client.sendRequest('_setVMTimelineFlags',
+        {'recordedStreams': ['Compiler', 'Dart', 'Embedder', 'GC']}
+    );
+    await client.sendRequest('_clearVMTimeline');
+  }
+
+  Future<String> stopTracing(AndroidApk apk, int observatoryPort, String outPath) async {
+    rpc.Client client;
+    try {
+      client = await _connectToObservatory(observatoryPort);
+    } catch (e) {
+      printError('Error connecting to observatory: $e');
+      return null;
     }
-    printError('No trace file detected. '
-        'Did you remember to start the trace before stopping it?');
-    return null;
+
+    await client.sendRequest('_setVMTimelineFlags', {'recordedStreams': '[]'});
+
+    String localPath;
+    if (outPath != null) {
+      localPath = outPath;
+    } else {
+      String timestamp = new DateFormat('yyyy_MM_dd_hh_mm_ss').format(new DateTime.now());
```

Comment:
If it's not critical that the file includes the current date and time, we could write the file out as `trace_xx.json` here. We use that format for crash reports and for screenshots. There's a utility method (https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/utils.dart#L44) to create a unique file w/ that name pattern.